lightning/src/offers/static_invoice.rs

```diff
@@ -0,0 +1,1173 @@
+	message_paths: Vec<BlindedPath>,
+}
+
+/// Builds a [`StaticInvoice`] from an [`Offer`].
+///
+/// See [module-level documentation] for usage.
+///
+/// [`Offer`]: crate::offers::offer::Offer
+/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+/// [`Bolt12Invoice`]: crate::offers::invoice::Bolt12Invoice
+/// This is not exported to bindings users as builder patterns don't map outside of move semantics.
+pub struct StaticInvoiceBuilder<'a> {
+	offer_bytes: &'a Vec<u8>,
+	invoice: InvoiceContents,
+	keys: Keypair,
+}
+
+impl<'a> StaticInvoiceBuilder<'a> {
+	/// Initialize a [`StaticInvoiceBuilder`] from the given [`Offer`].
+	///
+	/// Unless [`StaticInvoiceBuilder::relative_expiry`] is set, the invoice will expire 24 hours
+	/// after `created_at`.
+	pub fn for_offer_using_keys<T: secp256k1::Signing>(
+		offer: &'a Offer, payment_paths: Vec<(BlindedPayInfo, BlindedPath)>,
+		message_paths: Vec<BlindedPath>, created_at: Duration, expanded_key: &ExpandedKey,
+		secp_ctx: &Secp256k1<T>,
+	) -> Result<Self, Bolt12SemanticError> {
+		if offer.chains().len() > 1 {
+			return Err(Bolt12SemanticError::UnexpectedChain);
+		}
+
+		if payment_paths.is_empty() || message_paths.is_empty() || offer.paths().is_empty() {
+			return Err(Bolt12SemanticError::MissingPaths);
+		}
+
+		let offer_signing_pubkey =
+			offer.signing_pubkey().ok_or(Bolt12SemanticError::MissingSigningPubkey)?;
+
+		let keys = offer
+			.verify(&expanded_key, &secp_ctx)
+			.map_err(|()| Bolt12SemanticError::InvalidMetadata)?
+			.1
+			.ok_or(Bolt12SemanticError::MissingSigningPubkey)?;
+
+		let signing_pubkey = keys.public_key();
+		if signing_pubkey != offer_signing_pubkey {
+			return Err(Bolt12SemanticError::InvalidSigningPubkey);
+		}
+
+		let invoice =
+			InvoiceContents::new(offer, payment_paths, message_paths, created_at, signing_pubkey);
+
+		Ok(Self { offer_bytes: &offer.bytes, invoice, keys })
+	}
+
+	/// Builds a signed [`StaticInvoice`] after checking for valid semantics.
+	pub fn build_and_sign<T: secp256k1::Signing>(
+		self, secp_ctx: &Secp256k1<T>,
+	) -> Result<StaticInvoice, Bolt12SemanticError> {
+		#[cfg(feature = "std")]
+		{
+			if self.invoice.is_offer_expired() {
+				return Err(Bolt12SemanticError::AlreadyExpired);
+			}
+		}
+
+		#[cfg(not(feature = "std"))]
+		{
+			if self.invoice.is_offer_expired_no_std(self.invoice.created_at()) {
+				return Err(Bolt12SemanticError::AlreadyExpired);
+			}
+		}
+
+		let Self { offer_bytes, invoice, keys } = self;
+		let unsigned_invoice = UnsignedStaticInvoice::new(&offer_bytes, invoice);
+		let invoice = unsigned_invoice
+			.sign(|message: &UnsignedStaticInvoice| {
+				Ok(secp_ctx.sign_schnorr_no_aux_rand(message.tagged_hash.as_digest(), &keys))
+			})
+			.unwrap();
+		Ok(invoice)
+	}
+
+	invoice_builder_methods_common!(self, Self, self.invoice, Self, self, S, StaticInvoice, mut);
+}
+
+/// A semantically valid [`StaticInvoice`] that hasn't been signed.
+pub struct UnsignedStaticInvoice {
```

Comment:
Looks like this only needs to be `pub` if exposed in an `InvoiceBuilder::build` method. Should we go ahead and add a version that works with an `Offer` not using derived keys? Could leave as a follow-up, too.

Reply:
I'm good to add that in follow-up!